policy/release/slsa_source_correlated.rego

```diff
@@ -0,0 +1,130 @@
+# description: >-
+#   SLSA v1 verification model states:
+#
+#   "...artifacts are verified to ensure they meet the producer defined
+#   expectations of where the package source code was retrieved from..."
+#
+#   This package correlates the provided source code reference with the source
+#   code referenced in the attestation.
+#
+package policy.release.slsa_source_correlated
+
+import future.keywords.contains
+import future.keywords.if
+import future.keywords.in
+
+import data.lib
+
+# METADATA
+# title: Expected source code reference not provided
+# description: >-
+#   Warn if the expected source code reference is not provided.
+# custom:
+#   short_name: missing_expected_source_code_reference
+#   failure_msg: Expected source code reference was not provided for verification
+# solution: >-
+#   Provide the expected source code reference in inputs.
+warn contains result if {
+	source := object.get(input, ["image", "source"], {})
+	count(source) == 0
+	result := lib.result_helper(rego.metadata.chain(), [])
+}
+
+# METADATA
+# title: Source reference material not provided
+# description: >-
+#   Attested materials contain source reference.
+# custom:
+#   short_name: material_code_reference
+#   failure_msg: The attested material contains no source code reference
+#   solution: >-
+#     Check that the attestation creation process includes the source code reference
+#     in the predicate.materials for SLSA Provenance v0.2, or in
+#     predicate.buildDefinition.resolvedDependencies for SLSA Provenance v1.0
+#     attestations. Check that the SCM prefix is  the list of the supported SCM
+#     types in rule data (`supported_vcs` key).
+#   collections:
+#   - minimal
+#   - slsa1
+#   - slsa2
+#   - slsa3
+#   - redhat
+#   depends_on:
+#   - attestation_type.known_attestation_type
+#
+deny contains result if {
+	count(_source_references) == 0
+
+	result := lib.result_helper(rego.metadata.chain(), [])
+}
+
+# METADATA
+# title: Provided source code reference is the one that is being attested
+# description: >-
+#   Verify that the provided source code reference is the one being attested.
+# custom:
+#   short_name: expected_source_code_reference
+#   failure_msg: The expected source code reference %q is not the attested one %q
+#   solution: >-
+#     The source code reference in the attestation doesn't match the expected and
+#     provided source code reference. Make sure that the provided source code
+#     reference is correct, and if it is make sure that the build process is
+#     configured to retrieve the source code from the appropriate source code
+#     repository. Make sure that the source code reference is pointing to a
+#     explicit revision not to a symbolic identifier, e.g. a branch or tag name.
+#   collections:
+#   - minimal
+#   - slsa1
+#   - slsa2
+#   - slsa3
+#   - redhat
+#   depends_on:
+#   - attestation_type.known_attestation_type
+#
+deny contains result if {
+	some scm_type in object.keys(input.image.source)
```

Comment:
Nit: You could do this instead:

```rego
some scm_type, scm_info in input.image.source
...
expected_scm_uri := sprintf("%s+%s", [scm_type, scm_info.url])
```

See https://www.openpolicyagent.org/docs/latest/policy-reference/#objects-1